### PR TITLE
Datastore contention

### PR DESCRIPTION
**Pull request by wszarmach**
* Retry mechanism is based off the google example. Appears to work correctly.
* I don't like the solution to prevent transfer description exceeding 500 chars... but its past 2am and I cant think of anything better :disappointed: 
* Had to make a change to FormatUtil test because it stopped working for me.
* ActivityTypeConsistency check :grinning:

### CODE REVIEW

File: schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/model/consistency/ActivityTypeConsistency.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+@SuppressWarnings("unused")
+public class ActivityTypeConsistency implements EntityConsistency<ActivityType> {
+    @BeforeDelete(entityClass = ActivityType.class)
+    public void ensureActivityTypeHasNoActivities(Key id) throws InconsistentModelStateException {
+        List<Activity> activities = new ActivityDao().getByActivityTypeId(id);
+        if (!activities.isEmpty()) {
+            throw new InconsistentModelStateException("Cannot delete activity type with activities! " +
+                    "id=" + id + " (" + activities.size() + " activities).");
+        }
+    }
```
Comment (krico) [inline schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/model/consistency/ActivityTypeConsistency.java]: Cool!  You used it :-)

File: schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/spi/ActivityEndpoint.java [hunk 2]
```diff
@@ -114,41 +110,53 @@ public ActivityType updateActivityType(User caller, @Named("id") Key id, Activit
...
+
+        try {
+            return TransactionOperator.execute(new ModelOperation<ActivityType>() {
+                @Override
+                public ActivityType execute(Transaction tx) throws ModelException {
+                    activityTypeDao.save(activityType, organizationId);
+                    tx.commit();
+                    return activityType;
+                }
+            });
```
Comment (krico) [inline schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/spi/ActivityEndpoint.java]: Dude, you even used it on this same commit! :-)

File: schedule/schedule-appengine/src/test/java/com/jasify/schedule/appengine/util/FormatUtilTest.java [hunk 3]
```diff
@@ -17,6 +19,20 @@
 
 public class FormatUtilTest {
 
+
+    @Before
+    public void before() {
+        // Your wondering why... well initializeDatastore sets the default timezone to UTC.
+        // FormatUtil.toString is called indirectly by other tests and the SimpleDateFormat is initialized with default timezone
+        // By the time we get here its UTC but your pc default timezone is probably not UTC so the tests will fail
+        TestHelper.initializeDatastore();
```
Comment (krico) [inline schedule/schedule-appengine/src/test/java/com/jasify/schedule/appengine/util/FormatUtilTest.java]: I think this is the wrong fix for the case, relying on the default timezone to be UTC is probably not correct.  I can't understand what failed here so I might be wrong.

Comment (wszarmach) [inline schedule/schedule-appengine/src/test/java/com/jasify/schedule/appengine/util/FormatUtilTest.java]: The issue is that when TestHelper.initializeDatastore() is called the following code ends up being called
```
    public final LocalServiceTestHelper setUp() {
        this.originalDefaultTimeZone = TimeZone.getDefault();
        TimeZone.setDefault(this.timeZone);
```
When TestHelper.cleanupDatastore() is called the original timezone is restored.
START_FORMAT and FINISH_FORMAT are static and at initialisation they get a Calendar object which has the following initialisation
```
    private void initializeCalendar(Locale loc) {
        if (calendar == null) {
            assert loc != null;
            // The format object must be constructed using the symbols for this zone.
            // However, the calendar should use the current default TimeZone.
            // If this is not contained in the locale zone strings, then the zone
            // will be formatted using generic GMT+/-H:MM nomenclature.
            calendar = Calendar.getInstance(TimeZone.getDefault(), loc);
        }
    }
```
One of the FormatUtil tests exercises these static variables with a calendar object that has the default timezone.
```
    @Test
    public void testToStringActivity() throws Exception {
        Activity activity = new Activity();
        activity.setName("The Activity");
        GregorianCalendar calendar = new GregorianCalendar();
        calendar.setTimeZone(TimeZone.getDefault());
        calendar.set(1976, Calendar.JULY, 15, 22, 10);
        activity.setStart(calendar.getTime());
        calendar.add(Calendar.HOUR, 1);
        activity.setFinish(calendar.getTime());
        String formatted = FormatUtil.toString(activity);
        assertNotNull(formatted);
        assertEquals("The Activity [Thu, 15 Jul 22:10 - 23:10]", formatted);
    }
```
The issue is that if some other test that uses datastores runs earlier and triggers the initialisation of the START_FORMAT and FINISH_FORMAT variables they will be stuck at UTC but if FormatUtil does not change the default timezone to UTC its test will fail. I could change the test to also set the default timezone to UTC and then change back but this seemed simpler/easier/lasier

File: schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/model/balance/DefaultBalanceService.java [hunk 4]
```diff
@@ -161,8 +164,19 @@ public void payment(Payment payment) {
...
+        } catch (ModelException e) {
+            // TODO: This should not happen but if it ever does schedule a task to retry
+            log.error("Failed to make debit", e);
+        }
+
+        try {
+            applyTransferTransaction(transfer, false);
+        } catch (ModelException e) {
+            // TODO: This should not happen but if it ever does schedule a task to retry
+            log.error("Failed to make credit", e);
```
Comment (krico) [inline schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/model/balance/DefaultBalanceService.java]: I don't understand your TODO... Could you pls explain?

File: schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/spi/ActivityEndpoint.java [hunk 5]
```diff
@@ -91,64 +86,77 @@ public ActivityType getActivityType(User caller, @Named("id") Key id) throws Not
...
-            final ActivityType dbActivityType = activityTypeDao.get(activityType.getId());
-            final Key organizationId = dbActivityType.getOrganizationRef().getKey();
-            if (!StringUtils.equalsIgnoreCase(dbActivityType.getName(), activityType.getName())) {
-                if (activityTypeDao.exists(activityType.getName(), organizationId)) {
-                    throw new BadRequestException("ActivityType.name=" + activityType.getName() + ", Organization.id=" + organizationId);
+            return TransactionOperator.execute(new ModelOperation<ActivityType>() {
+                @Override
+                public ActivityType execute(Transaction tx) throws ModelException {
+                    ActivityType dbActivityType = activityTypeDao.get(id);
+                    BeanUtil.copyPropertiesExcluding(dbActivityType, activityType, "created", "modified", "id", "organizationRef");
```
Comment (krico) [inline schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/spi/ActivityEndpoint.java]: I raised #576 to get rid of commons-beanutils...

File: schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/model/balance/DefaultBalanceService.java (hunk 4 above)
Comment (wszarmach) [inline schedule/schedule-appengine/src/main/java/com/jasify/schedule/appengine/model/balance/DefaultBalanceService.java]: I think its possible that a datastore contention could occur when doing the transfer, for example due to account balance move in Jasify account under heavy traffic. If that were to happen than scheduling a task to retry later would resolve the issue. The todo is just a reminder for the future to look further into this.

**Pull request merged**